### DISCUSSION TITLE
Questions about `type` in defineing problems

### DISCUSSION THREAD

**Discussion by Da1sypetals** [argmin-rs/argmin#511] [General]
I do not exactly know how to conclude the question, so I am showing an example:
For example, I want to use the `Newton` solver, and I have to implement `Gradient` and `Hessian`.
I wonder what types can I use to be `type Gradient`? Am I allowed to use sparse matrices? Is this documented anywhere?
```rust
impl Gradient for Rosenbrock {
    type Gradient = /* what types are valid? */;
    // ... omitted
}
```
I browsed the docs and getting-started website but did not find anything related. 
Thanks a lot in advance❤️

**Comment by stefan-k** [comment]
You can use any type you want; however, this type is required to implement a number of traits which depend on the solver you are using. The easiest way is to use one of the types supported by the various math backends ([docs](https://docs.rs/argmin-math/latest/argmin_math/), [book](https://www.argmin-rs.org/book/using_argmin.html)), such as `Vec<f64>`, `Vec<f32>`, `ndarray::Array1<f64>`, `ndarray::Array1<f32>`, and the various vectors supported by `nalgebra`. 

If you choose to use your own type, you will have to implement the relevant traits on that type. For the Newton solver you can find the trait bounds listed [here](https://docs.rs/argmin/latest/argmin/solver/newton/struct.Newton.html#impl-Solver%3CO,+IterState%3CP,+G,+(),+H,+(),+F%3E%3E-for-Newton%3CF%3E). In particular you would have to implement `argmin_math::ArgminScaledSub` for your type and `argmin_math::ArgminInv` and `argmin_math::ArgminDot` for an associated matrix type `H`.
